map_reduce: prevent mapper exception from poisoning state

avikivity opened the pull request:
In map_reduce, if the mapper throws, we can end up with poisoned state. This is because the compiler can choose to move s->result before evaluating f.get(), so s->result gets broken (if the mapped type becomes invalid after move).

Fix by triggering the exception before the moving s->result, with a sequence point to force the ordering.

A unit test is added (which does fail with clang before the patch, though if the arguments are evaluated in reverse order, it can pass with a different compiler).

tgrabiec commented on include/seastar/core/map_reduce.hh:
```diff
@@ -191,7 +191,8 @@ map_reduce(Iterator begin, Iterator end, Mapper&& mapper, Initial initial, Reduc
             try {
-                s->result = s->reduce(std::move(s->result), f.get());
+                auto mapped_value = f.get(); // Extract exception before touching s->result
+                s->result = s->reduce(std::move(s->result), std::move(mapped_value));
```
Don't we have the same problem if reduce() throws?

avikivity replied:
Yes, good catch

avikivity commented:
v2:
 - rewrote to capture the exception in state
 - this enables handling exceptions from the reducer too
 - added unit test for throwing reducer (also fails before patch)
 - tightened test to ensure map_reduce actually throws

avikivity commented:
v3:
 -  ignore_ready_future if we already saw an exception
 - maintain task chain (`return std::move(ret)`) unconditionally
